ffi/capi/src/neo_datetime.rs

```diff
@@ -0,0 +1,505 @@
+        L,
+    }
+
+    #[diplomat::rust_link(icu::datetime::builder::FieldSetBuilder, Enum)]
+    pub struct DateTimeFieldSetBuilder {
+        pub length: DiplomatOption<DateTimeLength>,
+        pub date_fields: DiplomatOption<DateFields>,
+        pub time_precision: DiplomatOption<TimePrecision>,
+        pub zone_style: DiplomatOption<ZoneStyle>,
+        pub alignment: DiplomatOption<DateTimeAlignment>,
+        pub year_style: DiplomatOption<YearStyle>,
+    }
+
+    #[diplomat::opaque]
+    #[diplomat::rust_link(icu::datetime::DateTimeFormatter, Typedef)]
+    pub struct DateTimeFormatter(pub icu_datetime::DateTimeFormatter<CompositeDateTimeFieldSet>);
+
+    impl DateTimeFormatter {
+        #[diplomat::attr(all(supports = fallible_constructors, supports = named_constructors), named_constructor = "from_builder")]
+        #[diplomat::rust_link(icu::datetime::fieldsets::DateTimeFormatter::try_new, FnInStruct)]
+        #[diplomat::demo(default_constructor)]
+        #[cfg(feature = "compiled_data")]
+        pub fn create_from_builder(
+            locale: &Locale,
+            builder: DateTimeFieldSetBuilder,
+        ) -> Result<Box<DateTimeFormatter>, DateTimeFormatterBuildOrLoadError> {
+            let prefs = (&locale.0).into();
+            let builder = icu_datetime::fieldsets::builder::FieldSetBuilder::from(builder);
+            let options = builder.build_composite_datetime()?;
+            Ok(Box::new(DateTimeFormatter(
+                icu_datetime::DateTimeFormatter::try_new(prefs, options)?,
+            )))
+        }
+
+        #[diplomat::attr(all(supports = fallible_constructors, supports = named_constructors), named_constructor = "from_builder_with_provider")]
+        #[diplomat::rust_link(icu::datetime::fieldsets::DateTimeFormatter::try_new, FnInStruct)]
+        #[cfg(feature = "compiled_data")]
+        pub fn create_from_builder_with_provider(
+            provider: &DataProvider,
+            locale: &Locale,
+            builder: DateTimeFieldSetBuilder,
+        ) -> Result<Box<DateTimeFormatter>, DateTimeFormatterBuildOrLoadError> {
+            let prefs = (&locale.0).into();
+            let builder = icu_datetime::fieldsets::builder::FieldSetBuilder::from(builder);
+            let options = builder.build_composite_datetime()?;
+            Ok(Box::new(DateTimeFormatter(
+                icu_datetime::DateTimeFormatter::try_new_with_buffer_provider(
+                    provider.get()?,
+                    prefs,
+                    options,
+                )?,
+            )))
+        }
+
+        #[diplomat::attr(all(supports = fallible_constructors, supports = named_constructors), named_constructor = "dt")]
+        #[diplomat::rust_link(icu::datetime::fieldsets::DT, Struct)]
+        #[cfg(feature = "compiled_data")]
+        pub fn create_dt(
+            locale: &Locale,
+            length: DateTimeLength,
+            time_precision: TimePrecision,
+            alignment: DateTimeAlignment,
```

Comment:
should these be optional?

Reply:
This is exactly one of things this PR is intended to be a place to bikeshed. I posted several choices in https://github.com/unicode-org/icu4x/pull/6035#issuecomment-2614229124. Positional option arguments is another approach I didn't consider.

Reply:
The choices in the comment bunch together multiple axes: whether to use options bags or parameters, and whether the options are options. I don't think we should ever force a user to supply all arguments, so the three APIs we should discuss are single options bag, options bag per type, and option parameters.